File: Baitap1/Complete/Python/7.5.1.py
def printMove(fr, to):
    print('Move from ' + str(fr) + ' to ' + str(to))

def solveHanoiTowers(n, fr, to, spare):
    if n == 1:
        printMove(fr, to)
    else:
        solveHanoiTowers(n - 1, fr, spare, to)
        solveHanoiTowers(1, fr, to, spare)
        solveHanoiTowers(n - 1, spare, to, fr)

solveHanoiTowers(4, 'P1', 'P2', 'P3')

File: Baitap1/Complete/Java/7.5.1.java
public class Main {
    private static void printMove(String from, String to) {
        System.out.println("Move from " + from + " to " + to);
    }
    
    private static void solveHanoiTower(int n, String from, String to, String spare) {
        if (n == 1) {
            printMove(from, to);
        } else {
            solveHanoiTower(n - 1, from, spare, to);
            solveHanoiTower(1, from, to, spare);
            solveHanoiTower(n - 1, spare, to, from);
        }
    }
    public static void main(String[] args) {
        solveHanoiTower(4, "P1", "P2", "P3");
    }
}